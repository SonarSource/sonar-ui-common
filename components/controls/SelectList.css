/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

.select-list-container {
  min-width: 500px;
  box-sizing: border-box;
}

.select-list-control {
  margin-bottom: 10px;
  box-sizing: border-box;
}

.select-list-list-container {
  border: 1px solid #bfbfbf;
  box-sizing: border-box;
  height: 400px;
  overflow: auto;
}

.select-list-list-checkbox {
  display: flex !important;
  align-items: center;
}

.select-list-list-checkbox i {
  display: inline-block;
  vertical-align: middle;
  margin-right: 10px;
}

.select-list-list-disabled {
  cursor: not-allowed;
}

.select-list-list-disabled > a {
  pointer-events: none;
}

.select-list-list-item {
  display: inline-block;
  vertical-align: middle;
}
