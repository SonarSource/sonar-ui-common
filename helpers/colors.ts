/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
/* eslint-disable no-bitwise, no-mixed-operators */
export function stringToColor(str: string) {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    hash = str.charCodeAt(i) + ((hash << 5) - hash);
  }
  let color = '#';
  for (let i = 0; i < 3; i++) {
    const value = (hash >> (i * 8)) & 0xff;
    color += ('00' + value.toString(16)).substr(-2);
  }
  return color;
}

export function isDarkColor(color: string) {
  color = color.substr(1);
  if (color.length === 3) {
    // shortcut notation: #f90
    color = color[0] + color[0] + color[1] + color[1] + color[2] + color[2];
  }
  const rgb = parseInt(color.substr(1), 16);
  const r = (rgb >> 16) & 0xff;
  const g = (rgb >> 8) & 0xff;
  const b = (rgb >> 0) & 0xff;
  const luma = 0.2126 * r + 0.7152 * g + 0.0722 * b;
  return luma < 140;
}

export function getTextColor(background: string, dark = '#222', light = '#fff') {
  return isDarkColor(background) ? light : dark;
}
