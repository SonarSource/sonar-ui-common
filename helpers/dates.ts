/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as parse from 'date-fns/parse';

function pad(number: number) {
  if (number < 10) {
    return '0' + number.toString();
  }
  return number;
}

type ParsableDate = string | number | Date;

export function parseDate(rawDate: ParsableDate): Date {
  return parse(rawDate);
}

export function toShortNotSoISOString(rawDate: ParsableDate): string {
  const date = parseDate(rawDate);
  return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`;
}

export function toNotSoISOString(rawDate: ParsableDate): string {
  const date = parseDate(rawDate);
  return date.toISOString().replace(/\..+Z$/, '+0000');
}

export function isValidDate(date: Date): boolean {
  return !isNaN(date.getTime());
}
