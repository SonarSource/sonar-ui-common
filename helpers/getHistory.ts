/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { createHistory, History } from 'history';
import { useRouterHistory } from 'react-router';

let history: History;

function ensureHistory() {
  // eslint-disable-next-line react-hooks/rules-of-hooks
  history = useRouterHistory(createHistory)({
    basename: (window as any).baseUrl,
  });
  return history;
}

export default function getHistory() {
  return history ? history : ensureHistory();
}
