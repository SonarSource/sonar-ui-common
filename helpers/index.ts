/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

export * from './colors';
export * from './cookies';
export * from './csv';
export * from './dates';
export * from './getHistory';
export * from './handleRequiredAuthentication';
export * from './init';
export * from './keycodes';
export * from './l10n';
export * from './measures';
export * from './pages';
export * from './path';
export * from './query';
export * from './ratings';
export * from './request';
export * from './scrolling';
export * from './search';
export {
  get as getFromLocalStorage,
  remove as removeFromLocalStorage,
  save as saveInLocalStorage,
} from './storage';
export * from './strings';
export * from './testUtils';
export * from './urls';
