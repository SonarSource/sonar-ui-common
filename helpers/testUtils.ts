/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { ReactWrapper, ShallowWrapper } from 'enzyme';

export function mockEvent(overrides = {}) {
  return {
    target: { blur() {} },
    currentTarget: { blur() {} },
    preventDefault() {},
    stopPropagation() {},
    ...overrides
  } as any;
}

export function click(element: ShallowWrapper | ReactWrapper, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onClick')();
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('click', mockEvent(event));
  }
}

export function clickOutside(event = {}): void {
  const dispatchedEvent = new MouseEvent('click', event);
  window.dispatchEvent(dispatchedEvent);
}

export function submit(element: ShallowWrapper | ReactWrapper): void {
  element.simulate('submit', {
    preventDefault() {}
  });
}

export function change(element: ShallowWrapper | ReactWrapper, value: string, event = {}): void {
  // `type()` returns a component constructor for a composite element and string for DOM nodes
  if (typeof element.type() === 'function') {
    element.prop<Function>('onChange')(value);
    // TODO find out if `root` is a public api
    // https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ReactWrapper.js#L109
    (element as any).root().update();
  } else {
    element.simulate('change', {
      target: { value },
      currentTarget: { value },
      ...event
    });
  }
}

export const KEYCODE_MAP: { [keycode: number]: string } = {
  13: 'enter',
  37: 'left',
  38: 'up',
  39: 'right',
  40: 'down'
};

export function keydown(key: number | string): void {
  let keyCode;
  if (typeof key === 'number') {
    keyCode = key;
  } else {
    // eslint-disable-next-line no-console
    console.warn('Using strings in keydown() is deprecated. Consider using the KeyCodes enum.');
    const mapped = Object.entries(KEYCODE_MAP).find(([_, value]) => value === key);
    if (!mapped) {
      throw new Error(`Cannot map key "${key}" to a keyCode!`);
    }
    keyCode = mapped[0];
  }

  const event = new KeyboardEvent('keydown', { keyCode, which: keyCode } as KeyboardEventInit);
  document.dispatchEvent(event);
}

export function elementKeydown(element: ShallowWrapper, keyCode: number): void {
  const event = {
    currentTarget: { element },
    keyCode,
    preventDefault() {}
  };

  if (typeof element.type() === 'string') {
    // `type()` is string for native dom elements
    element.simulate('keydown', event);
  } else {
    element.prop<Function>('onKeyDown')(event);
  }
}

export function resizeWindowTo(width?: number, height?: number) {
  // `document.documentElement.clientHeight/clientWidth` are getters by default,
  // so we need to redefine them. Pass `configurable: true` to allow to redefine
  // the properties multiple times.
  if (width) {
    Object.defineProperty(document.documentElement, 'clientWidth', {
      configurable: true,
      value: width
    });
  }
  if (height) {
    Object.defineProperty(document.documentElement, 'clientHeight', {
      configurable: true,
      value: height
    });
  }

  const resizeEvent = new Event('resize');
  window.dispatchEvent(resizeEvent);
}

export function scrollTo({ left = 0, top = 0 }) {
  Object.defineProperty(window, 'pageYOffset', { value: top });
  Object.defineProperty(window, 'pageXOffset', { value: left });
  const resizeEvent = new Event('scroll');
  window.dispatchEvent(resizeEvent);
}

export function setNodeRect({ width = 50, height = 50, left = 0, top = 0 }) {
  const { findDOMNode } = require('react-dom');
  const element = document.createElement('div');
  Object.defineProperty(element, 'getBoundingClientRect', {
    value: () => ({ width, height, left, top })
  });
  findDOMNode.mockReturnValue(element);
}

export function doAsync(fn?: Function): Promise<void> {
  return new Promise(resolve => {
    setImmediate(() => {
      if (fn) {
        fn();
      }
      resolve();
    });
  });
}

export async function waitAndUpdate(wrapper: ShallowWrapper<any, any> | ReactWrapper<any, any>) {
  await new Promise(setImmediate);
  wrapper.update();
}
