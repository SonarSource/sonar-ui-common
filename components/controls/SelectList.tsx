/*
 * Sonar UI Common
 * Copyright (C) 2019-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import SelectListListContainer from './SelectListListContainer';
import SearchBox from './SearchBox';
import RadioToggle from './RadioToggle';
import ListFooter from './ListFooter';
import { translate } from '../../helpers/l10n';
import './SelectList.css';

export enum SelectListFilter {
  All = 'all',
  Selected = 'selected',
  Unselected = 'deselected'
}

interface Props {
  allowBulkSelection?: boolean;
  elements: string[];
  elementsTotalCount?: number;
  disabledElements?: string[];
  labelSelected?: string;
  labelUnselected?: string;
  labelAll?: string;
  needToReload?: boolean;
  onSearch: (searchParams: SelectListSearchParams) => Promise<void>;
  onSelect: (element: string) => Promise<void>;
  onUnselect: (element: string) => Promise<void>;
  pageSize?: number;
  readOnly?: boolean;
  renderElement: (element: string) => React.ReactNode;
  selectedElements: string[];
  withPaging?: boolean;
}

export interface SelectListSearchParams {
  filter: SelectListFilter;
  page?: number;
  pageSize?: number;
  query: string;
}

interface State {
  lastSearchParams: SelectListSearchParams;
  loading: boolean;
}

const DEFAULT_PAGE_SIZE = 100;

export default class SelectList extends React.PureComponent<Props, State> {
  mounted = false;

  constructor(props: Props) {
    super(props);

    this.state = {
      lastSearchParams: {
        filter: SelectListFilter.Selected,
        page: 1,
        pageSize: props.pageSize ? props.pageSize : DEFAULT_PAGE_SIZE,
        query: ''
      },
      loading: false
    };
  }

  componentDidMount() {
    this.mounted = true;
    this.search({});
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  stopLoading = () => {
    if (this.mounted) {
      this.setState({ loading: false });
    }
  };

  getFilter = () =>
    this.state.lastSearchParams.query === ''
      ? this.state.lastSearchParams.filter
      : SelectListFilter.All;

  search = (searchParams: Partial<SelectListSearchParams>) =>
    this.setState(
      prevState => ({
        loading: true,
        lastSearchParams: { ...prevState.lastSearchParams, ...searchParams }
      }),
      () =>
        this.props
          .onSearch({
            filter: this.getFilter(),
            page: this.props.withPaging ? this.state.lastSearchParams.page : undefined,
            pageSize: this.props.withPaging ? this.state.lastSearchParams.pageSize : undefined,
            query: this.state.lastSearchParams.query
          })
          .then(this.stopLoading)
          .catch(this.stopLoading)
    );

  changeFilter = (filter: SelectListFilter) => this.search({ filter, page: 1 });

  handleQueryChange = (query: string) => this.search({ page: 1, query });

  onLoadMore = () =>
    this.search({
      page:
        this.state.lastSearchParams.page != null ? this.state.lastSearchParams.page + 1 : undefined
    });

  onReload = () => this.search({ page: 1 });

  render() {
    const {
      labelSelected = translate('selected'),
      labelUnselected = translate('unselected'),
      labelAll = translate('all')
    } = this.props;
    const { filter } = this.state.lastSearchParams;

    const disabled = this.state.lastSearchParams.query !== '';

    return (
      <div className="select-list">
        <div className="display-flex-center">
          <RadioToggle
            className="select-list-filter spacer-right"
            name="filter"
            onCheck={this.changeFilter}
            options={[
              { disabled, label: labelSelected, value: SelectListFilter.Selected },
              { disabled, label: labelUnselected, value: SelectListFilter.Unselected },
              { disabled, label: labelAll, value: SelectListFilter.All }
            ]}
            value={filter}
          />
          <SearchBox
            autoFocus={true}
            loading={this.state.loading}
            onChange={this.handleQueryChange}
            placeholder={translate('search_verb')}
            value={this.state.lastSearchParams.query}
          />
        </div>
        <SelectListListContainer
          allowBulkSelection={this.props.allowBulkSelection}
          disabledElements={this.props.disabledElements || []}
          elements={this.props.elements}
          filter={this.getFilter()}
          onSelect={this.props.onSelect}
          onUnselect={this.props.onUnselect}
          readOnly={this.props.readOnly}
          renderElement={this.props.renderElement}
          selectedElements={this.props.selectedElements}
        />
        {!!this.props.elementsTotalCount && (
          <ListFooter
            count={this.props.elements.length}
            loadMore={this.onLoadMore}
            needReload={this.props.needToReload}
            reload={this.onReload}
            total={this.props.elementsTotalCount}
          />
        )}
      </div>
    );
  }
}
