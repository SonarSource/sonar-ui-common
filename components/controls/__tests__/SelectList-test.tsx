/*
 * Sonar UI Common
 * Copyright (C) 2019-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { waitAndUpdate } from '../../../helpers/testUtils';
import SelectList, { SelectListFilter } from '../SelectList';

const elements = ['foo', 'bar', 'baz'];
const selectedElements = [elements[0]];
const disabledElements = [elements[1]];

it('should display properly with basics features', async () => {
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);
  expect(wrapper.instance().mounted).toBe(true);

  expect(wrapper).toMatchSnapshot();

  wrapper.instance().componentWillUnmount();
  expect(wrapper.instance().mounted).toBe(false);
});

it('should display properly with advanced features', async () => {
  const wrapper = shallowRender({
    allowBulkSelection: true,
    elementsTotalCount: 125,
    pageSize: 10,
    readOnly: true,
    withPaging: true
  });
  await waitAndUpdate(wrapper);

  expect(wrapper).toMatchSnapshot();
});

it('should display a loader when searching', async () => {
  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);
  expect(wrapper.state().loading).toBe(false);

  wrapper.instance().search({});
  expect(wrapper.state().loading).toBe(true);
  expect(wrapper).toMatchSnapshot();

  await waitAndUpdate(wrapper);
  expect(wrapper.state().loading).toBe(false);
});

it('should cancel filter selection when search is active', async () => {
  const spy = jest.fn().mockResolvedValue({});
  const wrapper = shallowRender({ onSearch: spy });
  wrapper.instance().changeFilter(SelectListFilter.Unselected);
  await waitAndUpdate(wrapper);

  expect(spy).toHaveBeenCalledWith({
    query: '',
    filter: SelectListFilter.Unselected,
    page: undefined,
    pageSize: undefined
  });
  expect(wrapper).toMatchSnapshot();

  const query = 'test';
  wrapper.instance().handleQueryChange(query);
  expect(spy).toHaveBeenCalledWith({
    query,
    filter: SelectListFilter.All,
    page: undefined,
    pageSize: undefined
  });

  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();

  wrapper.instance().handleQueryChange('');
  expect(spy).toHaveBeenCalledWith({
    query: '',
    filter: SelectListFilter.Unselected,
    page: undefined,
    pageSize: undefined
  });

  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
});

it('should display pagination element properly and call search method with correct parameters', () => {
  const spy = jest.fn().mockResolvedValue({});
  const wrapper = shallowRender({ elementsTotalCount: 100, onSearch: spy, withPaging: true });
  expect(wrapper).toMatchSnapshot();
  expect(spy).toHaveBeenCalledWith({
    query: '',
    filter: SelectListFilter.Selected,
    page: 1,
    pageSize: 100
  }); // Basic default call

  wrapper.instance().onLoadMore();
  expect(spy).toHaveBeenCalledWith({
    query: '',
    filter: SelectListFilter.Selected,
    page: 2,
    pageSize: 100
  }); // Load more call

  wrapper.instance().onReload();
  expect(spy).toHaveBeenCalledWith({
    query: '',
    filter: SelectListFilter.Selected,
    page: 1,
    pageSize: 100
  }); // Reload call

  wrapper.setProps({ needToReload: true });
  expect(wrapper).toMatchSnapshot();
});

function shallowRender(props: Partial<SelectList['props']> = {}) {
  return shallow<SelectList>(
    <SelectList
      disabledElements={disabledElements}
      elements={elements}
      onSearch={jest.fn(() => Promise.resolve())}
      onSelect={jest.fn(() => Promise.resolve())}
      onUnselect={jest.fn(() => Promise.resolve())}
      renderElement={(foo: string) => foo}
      selectedElements={selectedElements}
      {...props}
    />
  );
}
