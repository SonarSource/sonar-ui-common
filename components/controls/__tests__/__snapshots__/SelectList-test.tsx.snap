// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should cancel filter selection when search is active 1`] = `
<div
  className="select-list"
>
  <div
    className="display-flex-center"
  >
    <RadioToggle
      className="select-list-filter spacer-right"
      disabled={false}
      name="filter"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "disabled": false,
            "label": "selected",
            "value": "selected",
          },
          Object {
            "disabled": false,
            "label": "unselected",
            "value": "deselected",
          },
          Object {
            "disabled": false,
            "label": "all",
            "value": "all",
          },
        ]
      }
      value="deselected"
    />
    <SearchBox
      autoFocus={true}
      loading={false}
      onChange={[Function]}
      placeholder="search_verb"
      value=""
    />
  </div>
  <SelectListListContainer
    disabledElements={
      Array [
        "bar",
      ]
    }
    elements={
      Array [
        "foo",
        "bar",
        "baz",
      ]
    }
    filter="deselected"
    onSelect={[MockFunction]}
    onUnselect={[MockFunction]}
    renderElement={[Function]}
    selectedElements={
      Array [
        "foo",
      ]
    }
  />
</div>
`;

exports[`should cancel filter selection when search is active 2`] = `
<div
  className="select-list"
>
  <div
    className="display-flex-center"
  >
    <RadioToggle
      className="select-list-filter spacer-right"
      disabled={false}
      name="filter"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "disabled": true,
            "label": "selected",
            "value": "selected",
          },
          Object {
            "disabled": true,
            "label": "unselected",
            "value": "deselected",
          },
          Object {
            "disabled": true,
            "label": "all",
            "value": "all",
          },
        ]
      }
      value="deselected"
    />
    <SearchBox
      autoFocus={true}
      loading={false}
      onChange={[Function]}
      placeholder="search_verb"
      value="test"
    />
  </div>
  <SelectListListContainer
    disabledElements={
      Array [
        "bar",
      ]
    }
    elements={
      Array [
        "foo",
        "bar",
        "baz",
      ]
    }
    filter="all"
    onSelect={[MockFunction]}
    onUnselect={[MockFunction]}
    renderElement={[Function]}
    selectedElements={
      Array [
        "foo",
      ]
    }
  />
</div>
`;

exports[`should cancel filter selection when search is active 3`] = `
<div
  className="select-list"
>
  <div
    className="display-flex-center"
  >
    <RadioToggle
      className="select-list-filter spacer-right"
      disabled={false}
      name="filter"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "disabled": false,
            "label": "selected",
            "value": "selected",
          },
          Object {
            "disabled": false,
            "label": "unselected",
            "value": "deselected",
          },
          Object {
            "disabled": false,
            "label": "all",
            "value": "all",
          },
        ]
      }
      value="deselected"
    />
    <SearchBox
      autoFocus={true}
      loading={false}
      onChange={[Function]}
      placeholder="search_verb"
      value=""
    />
  </div>
  <SelectListListContainer
    disabledElements={
      Array [
        "bar",
      ]
    }
    elements={
      Array [
        "foo",
        "bar",
        "baz",
      ]
    }
    filter="deselected"
    onSelect={[MockFunction]}
    onUnselect={[MockFunction]}
    renderElement={[Function]}
    selectedElements={
      Array [
        "foo",
      ]
    }
  />
</div>
`;

exports[`should display a loader when searching 1`] = `
<div
  className="select-list"
>
  <div
    className="display-flex-center"
  >
    <RadioToggle
      className="select-list-filter spacer-right"
      disabled={false}
      name="filter"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "disabled": false,
            "label": "selected",
            "value": "selected",
          },
          Object {
            "disabled": false,
            "label": "unselected",
            "value": "deselected",
          },
          Object {
            "disabled": false,
            "label": "all",
            "value": "all",
          },
        ]
      }
      value="selected"
    />
    <SearchBox
      autoFocus={true}
      loading={true}
      onChange={[Function]}
      placeholder="search_verb"
      value=""
    />
  </div>
  <SelectListListContainer
    disabledElements={
      Array [
        "bar",
      ]
    }
    elements={
      Array [
        "foo",
        "bar",
        "baz",
      ]
    }
    filter="selected"
    onSelect={[MockFunction]}
    onUnselect={[MockFunction]}
    renderElement={[Function]}
    selectedElements={
      Array [
        "foo",
      ]
    }
  />
</div>
`;

exports[`should display pagination element properly and call search method with correct parameters 1`] = `
<div
  className="select-list"
>
  <div
    className="display-flex-center"
  >
    <RadioToggle
      className="select-list-filter spacer-right"
      disabled={false}
      name="filter"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "disabled": false,
            "label": "selected",
            "value": "selected",
          },
          Object {
            "disabled": false,
            "label": "unselected",
            "value": "deselected",
          },
          Object {
            "disabled": false,
            "label": "all",
            "value": "all",
          },
        ]
      }
      value="selected"
    />
    <SearchBox
      autoFocus={true}
      loading={true}
      onChange={[Function]}
      placeholder="search_verb"
      value=""
    />
  </div>
  <SelectListListContainer
    disabledElements={
      Array [
        "bar",
      ]
    }
    elements={
      Array [
        "foo",
        "bar",
        "baz",
      ]
    }
    filter="selected"
    onSelect={[MockFunction]}
    onUnselect={[MockFunction]}
    renderElement={[Function]}
    selectedElements={
      Array [
        "foo",
      ]
    }
  />
  <ListFooter
    count={3}
    loadMore={[Function]}
    reload={[Function]}
    total={100}
  />
</div>
`;

exports[`should display pagination element properly and call search method with correct parameters 2`] = `
<div
  className="select-list"
>
  <div
    className="display-flex-center"
  >
    <RadioToggle
      className="select-list-filter spacer-right"
      disabled={false}
      name="filter"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "disabled": false,
            "label": "selected",
            "value": "selected",
          },
          Object {
            "disabled": false,
            "label": "unselected",
            "value": "deselected",
          },
          Object {
            "disabled": false,
            "label": "all",
            "value": "all",
          },
        ]
      }
      value="selected"
    />
    <SearchBox
      autoFocus={true}
      loading={true}
      onChange={[Function]}
      placeholder="search_verb"
      value=""
    />
  </div>
  <SelectListListContainer
    disabledElements={
      Array [
        "bar",
      ]
    }
    elements={
      Array [
        "foo",
        "bar",
        "baz",
      ]
    }
    filter="selected"
    onSelect={[MockFunction]}
    onUnselect={[MockFunction]}
    renderElement={[Function]}
    selectedElements={
      Array [
        "foo",
      ]
    }
  />
  <ListFooter
    count={3}
    loadMore={[Function]}
    needReload={true}
    reload={[Function]}
    total={100}
  />
</div>
`;

exports[`should display properly with advanced features 1`] = `
<div
  className="select-list"
>
  <div
    className="display-flex-center"
  >
    <RadioToggle
      className="select-list-filter spacer-right"
      disabled={false}
      name="filter"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "disabled": false,
            "label": "selected",
            "value": "selected",
          },
          Object {
            "disabled": false,
            "label": "unselected",
            "value": "deselected",
          },
          Object {
            "disabled": false,
            "label": "all",
            "value": "all",
          },
        ]
      }
      value="selected"
    />
    <SearchBox
      autoFocus={true}
      loading={false}
      onChange={[Function]}
      placeholder="search_verb"
      value=""
    />
  </div>
  <SelectListListContainer
    allowBulkSelection={true}
    disabledElements={
      Array [
        "bar",
      ]
    }
    elements={
      Array [
        "foo",
        "bar",
        "baz",
      ]
    }
    filter="selected"
    onSelect={[MockFunction]}
    onUnselect={[MockFunction]}
    readOnly={true}
    renderElement={[Function]}
    selectedElements={
      Array [
        "foo",
      ]
    }
  />
  <ListFooter
    count={3}
    loadMore={[Function]}
    reload={[Function]}
    total={125}
  />
</div>
`;

exports[`should display properly with basics features 1`] = `
<div
  className="select-list"
>
  <div
    className="display-flex-center"
  >
    <RadioToggle
      className="select-list-filter spacer-right"
      disabled={false}
      name="filter"
      onCheck={[Function]}
      options={
        Array [
          Object {
            "disabled": false,
            "label": "selected",
            "value": "selected",
          },
          Object {
            "disabled": false,
            "label": "unselected",
            "value": "deselected",
          },
          Object {
            "disabled": false,
            "label": "all",
            "value": "all",
          },
        ]
      }
      value="selected"
    />
    <SearchBox
      autoFocus={true}
      loading={false}
      onChange={[Function]}
      placeholder="search_verb"
      value=""
    />
  </div>
  <SelectListListContainer
    disabledElements={
      Array [
        "bar",
      ]
    }
    elements={
      Array [
        "foo",
        "bar",
        "baz",
      ]
    }
    filter="selected"
    onSelect={[MockFunction]}
    onUnselect={[MockFunction]}
    renderElement={[Function]}
    selectedElements={
      Array [
        "foo",
      ]
    }
  />
</div>
`;
